File: company-misc/2303-optiver/2.py
""" 2. Risk Limits

Working at Optiver, you quickly get used to the idea of having millions of dollarsworth of orders "in the market" at any given time. But that doesn't mean we aren'taware of the risks! A misbehaving trading process could lose a lot of money withina few minutes. That is why every order gets checked against a set of risk limitsbefore being sent to the market

Your task is to write a simple version of such a limit-checking system. The systemprocesses orders, consisting of an instrument identifier, a timestamp (given inmilliseconds since midnight), an order volume (always a positive integer) and anorder price (always a positive double). You can assume that orders arrive inincreasing timestamp order.
The system also processes risk limits. For every instrument, the following limits aredefined:

If the value (volume * price) of the order is above a given threshold, your programshould output MAX VAL LIMIT followed by the instrument name, e.g. MAX VAL LIMIT ABC.
If the total volume of all orders in a ten-second period exceeds a certain threshold.your program should output MAX VOL 10S LIMIT and the instrument name.
If the total value of all orders in a one-second period would exceed a certainthreshold, your program should outout MAX VAL 1S LIMIT, again followed the byinstrument name.

For every order, your program should print only the first limit breach it encountersIf multiple limit breaches are present in a single order, they should be checked inthe order given above.

We can't place orders without having limits defined for them; if this happens, your program should output NO LIMITS and the instrument name. Also note that limits may change throughout the day, and the latest version received should be used fochecking any orders received afterwards.

## Function Description
Your task is to implement a class which provides the methods AddLimit andProcessOrder.

## Input Format For Custom Testing
The first line contains an integer, n, denoting the number of elements in ARRAY_NAME.
Each line i of the n subsequent lines (where 0<=i< n) contains a(n) DATA_TYPE describing ARRAY_NAME_i
 """
from collections import defaultdict, deque

class RiskLimitProcessor:
    def __init__(self):
        self.limits = {}
        self.instru2orders = defaultdict(list)
        # val10, val1, p10, p1 = 0, 0, 0, 0
        self.instru2vals = defaultdict(lambda : (0,0,0,0))
        self.instru2limit = defaultdict(lambda: False)
    
    def AddLimit(self, instrument, maxValue, maxVolume10Seconds, maxValue1Second):
        self.limits[instrument] = (maxValue, maxVolume10Seconds, maxValue1Second)

    def ProcessOrder(self, instrument, timestamp, volume, price):
        if self.instru2limit[instrument]:
            return
        timestamp /= 1000

        val10, val1, p10, p1 = self.instru2vals[instrument]
        orders = self.instru2orders[instrument]
        if instrument in self.limits:
            maxValue, maxVolume10Seconds, maxValue1Second = self.limits[instrument]
            if volume*price > maxValue:
                print(f"MAX_VAL_LIMIT {instrument}")
                self.instru2limit[instrument] = True
                return
            # p10, p1 = self.instru2pointer10[instrument], self.instru2pointer1[instrument]

            while p10<len(orders) and orders[p10][0] < timestamp-10:
                val10 -= orders[p10][1]*orders[p10][2]
                p10 += 1
            if val10 + volume*price > maxVolume10Seconds:
                print(f"MAX_VAL_10S_LIMIT {instrument}")
                self.instru2limit[instrument] = True
                return
            while p1<len(orders) and orders[p1][0] < timestamp-1:
                val1 -= orders[p1][1]*orders[p1][2]
                p1 += 1
            if val1 + volume*price > maxValue1Second:
                print(f"MAX_VAL_1S_LIMIT {instrument}")
                self.instru2limit[instrument] = True
                return
        orders.append((timestamp, volume, price))
        val10 += volume*price
        val1 += volume*price
        self.instru2vals[instrument] = (val10, val1, p10, p1)

        # print(f"NO_LIMITS {instrument}")



def main():
    riskLimitProcessor = RiskLimitProcessor()
    while True:
        try:
            inp = input()
        except EOFError:
            break
        if not inp: break
        action, instrument,a,b,c = inp.split()
        if action == "LIMIT":
            riskLimitProcessor.AddLimit(instrument, float(a), int(b), float(c))
        elif action == "ORDER":
            riskLimitProcessor.ProcessOrder(instrument, int(a), int(b), float(c))
        else:
            print("Malformed input!")
            return(-1)
    return 0


if __name__=="__main__":
    main()

File: interview-list/230315-optiver/2.cpp
#include <cassert>
#include <map>
#include <set>
#include <list>
#include <cmath>
#include <ctime>
#include <deque>
#include <queue>
#include <stack>
#include <string>
#include <bitset>
#include <cstdio>
#include <limits>
#include <vector>
#include <climits>
#include <cstring>
#include <cstdlib>
#include <fstream>
#include <numeric>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <unordered_map>
#include <iomanip>
#include <array>


// This is the class you need to implement! Feel free to add members, private methods etc, but don't change the public
// method signatures.
// class RiskLimitProcessor
// {
//     public:
//         void AddLimit(const std::string& instrument, double maxValue, int maxVolume10Seconds, double maxValue1Second)
//         {
//         }

//         void ProcessOrder(const std::string& instrument, uint64_t timestamp, int volume, double price)
//         {
//             std::cout << "NO_LIMITS XXX" << std::endl;
//         }
// };

using namespace std;

class RiskLimitProcessor {
private:
    unordered_map<string, tuple<double, int, double>> limits;
    unordered_map<string, vector<tuple<int, int, double>>> instru2orders;
    unordered_map<string, tuple<int, int, double>> instru2vals;
public:
    void AddLimit(string instrument, double maxValue, int maxVolume10Seconds, double maxValue1Second) {
        limits[instrument] = make_tuple(maxValue, maxVolume10Seconds, maxValue1Second);
    }

    void ProcessOrder(string instrument, int timestamp, int volume, double price) {
        timestamp /= 1000;
        auto [maxValue, maxVolume10Seconds, maxValue1Second] = limits[instrument];
        if (volume * price > maxValue) {
            cout << "MAX_VAL_1S_LIMIT " << instrument << endl;
            return;
        }
        auto [val10, val1, p10, p1] = instru2vals[instrument];
        auto& orders = instru2orders[instrument];
        while (p10 < orders.size() && get<0>(orders[p10]) < timestamp - 10) {
            val10 -= get<1>(orders[p10]) * get<2>(orders[p10]);
            p10 += 1;
        }
        if (val10 + volume * price > maxVolume10Seconds) {
            cout << "MAX_VAL_10S_LIMIT " << instrument << endl;
            return;
        }
        while (p1 < orders.size() && get<0>(orders[p1]) < timestamp - 1) {
            val1 -= get<1>(orders[p1]) * get<2>(orders[p1]);
            p1 += 1;
        }
        if (val1 + volume * price > maxValue1Second) {
            cout << "MAX_VAL_LIMIT " << instrument << endl;
            return;
        }
        orders.emplace_back(timestamp, volume, price);
        val10 += volume * price;
        val1 += volume * price;
        instru2vals[instrument] = make_tuple(val10, val1, p10, p1);
    }
};



int main()
{
    RiskLimitProcessor riskLimitProcessor;
    while(!std::cin.eof())
    {
        std::string action, instrument;
        std::cin >> action >> instrument;
        if (action.empty())
            break; // handle whitespace at end of input
        if (action == "LIMIT")
        {
            double maxValue;
            int maxVolume10Seconds;
            double maxValue1Second;
            std::cin >> maxValue >> maxVolume10Seconds >> maxValue1Second;
            riskLimitProcessor.AddLimit(instrument, maxValue, maxVolume10Seconds, maxValue1Second);
        }
        else if (action == "ORDER")
        {
            uint64_t timestamp;
            int volume;
            double price;
            std::cin >> timestamp >> volume >> price;
            riskLimitProcessor.ProcessOrder(instrument, timestamp, volume, price);
        }
        else
        {
            std::cerr << "Malformed input!\n";
            return -1;
        }
    }
    return 0;
}